Neutrino tables do not get created on app start when neutrino enabled

I ran our node this weekend with neutrino enabled, and once we synced headers and started trying to sync filters, i got an error saying that `cfheaders` and `cfilters` tables do not exist.

>[info] 2019-10-07T13:42:23 ERROR [P2P] Error when processing headers message
[info] org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (no such table: cfheaders)
[info]  at org.sqlite.core.DB.newSQLException(DB.java:941)
[info]  at org.sqlite.core.DB.newSQLException(DB.java:953)
[info]  at org.sqlite.core.DB.throwex(DB.java:918)
[info]  at org.sqlite.core.NativeDB.prepare_utf8(Native Method)
[info]  at org.sqlite.core.NativeDB.prepare(NativeDB.java:134)
[info]  at org.sqlite.core.DB.prepare(DB.java:257)
[info]  at org.sqlite.core.CorePreparedStatement.<init>(CorePreparedStatement.java:47)
[info]  at org.sqlite.jdbc3.JDBC3PreparedStatement.<init>(JDBC3PreparedStatement.java:30)
[info]  at org.sqlite.jdbc4.JDBC4PreparedStatement.<init>(JDBC4PreparedStatement.java:19)
[info]  at org.sqlite.jdbc4.JDBC4Connection.prepareStatement(JDBC4Connection.java:35)
[info]  at org.sqlite.jdbc3.JDBC3Connection.prepareStatement(JDBC3Connection.java:241)
[info]  at org.sqlite.jdbc3.JDBC3Connection.prepareStatement(JDBC3Connection.java:205)
[info]  at slick.jdbc.JdbcBackend$SessionDef.prepareStatement(JdbcBackend.scala:375)
[info]  at slick.jdbc.JdbcBackend$SessionDef.prepareStatement$(JdbcBackend.scala:365)
[info]  at slick.jdbc.JdbcBackend$BaseSession.prepareStatement(JdbcBackend.scala:489)
[info]  at slick.jdbc.StatementInvoker.results(StatementInvoker.scala:33)
[info]  at slick.jdbc.StatementInvoker.iteratorTo(StatementInvoker.scala:22)
[info]  at slick.jdbc.Invoker.first(Invoker.scala:30)
[info]  at slick.jdbc.Invoker.first$(Invoker.scala:29)
[info]  at slick.jdbc.StatementInvoker.first(StatementInvoker.scala:16)
[info]  at slick.jdbc.JdbcActionComponent$QueryActionExtensionMethodsImpl$$anon$3.run(JdbcActionComponent.scala:224)
[info]  at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
[info]  at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)
[info]  at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.$anonfun$run$4(DBIOAction.scala:533)
[info]  at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.$anonfun$run$4$adapted(DBIOAction.scala:533)
[info]  at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
[info]  at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
[info]  at scala.collection.AbstractIterable.foreach(Iterable.scala:904)
[info]  at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.run(DBIOAction.scala:533)
[info]  at slick.basic.BasicBackend$DatabaseDef$$anon$3.liftedTree1$1(BasicBackend.scala:276)
[info]  at slick.basic.BasicBackend$DatabaseDef$$anon$3.run(BasicBackend.scala:276)
[info]  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]  at java.lang.Thread.run(Thread.java:748)
[info] [ERROR] [10/07/2019 08:42:23.861] [bitcoin-s-akka.actor.default-dispatcher-2] [akka://bitcoin-s/user/P2PClient$-1570455732843] [SQLITE_ERROR] SQL error or missing database (no such table: cfheaders)
[info] org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (no such table: cfheaders)
[info]  at org.sqlite.core.DB.newSQLException(DB.java:941)
[info]  at org.sqlite.core.DB.newSQLException(DB.java:953)
[info]  at org.sqlite.core.DB.throwex(DB.java:918)
[info]  at org.sqlite.core.NativeDB.prepare_utf8(Native Method)
[info]  at org.sqlite.core.NativeDB.prepare(NativeDB.java:134)
[info]  at org.sqlite.core.DB.prepare(DB.java:257)
[info]  at org.sqlite.core.CorePreparedStatement.<init>(CorePreparedStatement.java:47)
[info]  at org.sqlite.jdbc3.JDBC3PreparedStatement.<init>(JDBC3PreparedStatement.java:30)
[info]  at org.sqlite.jdbc4.JDBC4PreparedStatement.<init>(JDBC4PreparedStatement.java:19)
[info]  at org.sqlite.jdbc4.JDBC4Connection.prepareStatement(JDBC4Connection.java:35)
[info]  at org.sqlite.jdbc3.JDBC3Connection.prepareStatement(JDBC3Connection.java:241)
[info]  at org.sqlite.jdbc3.JDBC3Connection.prepareStatement(JDBC3Connection.java:205)
[info]  at slick.jdbc.JdbcBackend$SessionDef.prepareStatement(JdbcBackend.scala:375)
[info]  at slick.jdbc.JdbcBackend$SessionDef.prepareStatement$(JdbcBackend.scala:365)
[info]  at slick.jdbc.JdbcBackend$BaseSession.prepareStatement(JdbcBackend.scala:489)
[info]  at slick.jdbc.StatementInvoker.results(StatementInvoker.scala:33)
[info]  at slick.jdbc.StatementInvoker.iteratorTo(StatementInvoker.scala:22)
[info]  at slick.jdbc.Invoker.first(Invoker.scala:30)
[info]  at slick.jdbc.Invoker.first$(Invoker.scala:29)
[info]  at slick.jdbc.StatementInvoker.first(StatementInvoker.scala:16)
[info]  at slick.jdbc.JdbcActionComponent$QueryActionExtensionMethodsImpl$$anon$3.run(JdbcActionComponent.scala:224)
[info]  at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
[info]  at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)
[info]  at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.$anonfun$run$4(DBIOAction.scala:533)
[info]  at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.$anonfun$run$4$adapted(DBIOAction.scala:533)
[info]  at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
[info]  at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
[info]  at scala.collection.AbstractIterable.foreach(Iterable.scala:904)
[info]  at slick.dbio.SynchronousDatabaseAction$FusedAndThenAction.run(DBIOAction.scala:533)
[info]  at slick.basic.BasicBackend$DatabaseDef$$anon$3.liftedTree1$1(BasicBackend.scala:276)
[info]  at slick.basic.BasicBackend$DatabaseDef$$anon$3.run(BasicBackend.scala:276)
[info]  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]  at java.lang.Thread.run(Thread.java:748)


We need to incorporate this check into the chain project inside of `ChainAppConfig`

https://github.com/bitcoin-s/bitcoin-s/blob/933f0fcfd0c8233114b3d8cb004e26d8237246ae/chain/src/main/scala/org/bitcoins/chain/config/ChainAppConfig.scala#L38-L82

In general, it doesn't look like we have a good way to make sure our database is in sync with app code. We do have flyway added as a plugin, but i'm not sure how to bundle that into a jar and have that manage schemas.

--- a/app/server/src/main/scala/org/bitcoins/server/Main.scala
+++ b/app/server/src/main/scala/org/bitcoins/server/Main.scala
@@ -17,13 +17,10 @@ import scala.concurrent.duration._
 import scala.concurrent.{Await, Future}
 
 object Main extends App {
-  implicit val conf = {
-    // val custom = ConfigFactory.parseString("bitcoin-s.network = testnet3")
-    BitcoinSAppConfig.fromDefaultDatadir()
-  }
+  implicit val conf = BitcoinSAppConfig.fromDefaultDatadir()
 
   private val logger = HttpLogger.getLogger(
-    conf.walletConf // doesn't matter which one we pass in
+    conf.nodeConf // doesn't matter which one we pass in
   )
 
   implicit val walletConf: WalletAppConfig = conf.walletConf
@@ -130,7 +127,7 @@ object Main extends App {
   }
 
   startFut.failed.foreach { err =>
-    logger.info(s"Error on server startup!", err)
+    logger.error(s"Error on server startup!", err)
   }
 
   private def parseInetSocketAddress(

--- a/chain/src/main/scala/org/bitcoins/chain/config/ChainAppConfig.scala
+++ b/chain/src/main/scala/org/bitcoins/chain/config/ChainAppConfig.scala
@@ -1,17 +1,14 @@
 package org.bitcoins.chain.config
 
+import java.nio.file.Path
+
 import com.typesafe.config.Config
 import org.bitcoins.chain.db.ChainDbManagement
-import org.bitcoins.db._
 import org.bitcoins.chain.models.{BlockHeaderDAO, BlockHeaderDbHelper}
 import org.bitcoins.core.util.FutureUtil
+import org.bitcoins.db._
 
-import scala.concurrent.ExecutionContext
-import scala.concurrent.Future
-import scala.concurrent.Promise
-import scala.util.Success
-import scala.util.Failure
-import java.nio.file.Path
+import scala.concurrent.{ExecutionContext, Future}
 
 /** Configuration for the Bitcoin-S chain verification module
   * @param directory The data directory of the module
@@ -38,20 +35,17 @@ case class ChainAppConfig(
   def isInitialized()(implicit ec: ExecutionContext): Future[Boolean] = {
     val bhDAO =
       BlockHeaderDAO()(ec = implicitly[ExecutionContext], appConfig = this)
-    val p = Promise[Boolean]()
     val isDefinedOptF = {
       bhDAO.read(chain.genesisBlock.blockHeader.hashBE).map(_.isDefined)
     }
-    isDefinedOptF.onComplete {
-      case Success(bool) =>
-        logger.debug(s"Chain project is initialized")
-        p.success(bool)
-      case Failure(_) =>
-        logger.info(s"Chain project is not initialized")
-        p.success(false)
+    isDefinedOptF.foreach { _ =>
+      logger.debug(s"Chain project is initialized")
     }
-
-    p.future
+    isDefinedOptF.recover {
+      case e: Throwable =>
+        logger.info(s"Chain project is not initialized")
+        false
+    }
   }
 
   /** Initializes our chain project if it is needed
@@ -59,12 +53,14 @@ case class ChainAppConfig(
     * and inserts preliminary data like the genesis block header
     * */
   override def initialize()(implicit ec: ExecutionContext): Future[Unit] = {
-    val isInitF = isInitialized()
+    val createdF = ChainDbManagement.createAll()(this, ec)
+    val isInitF = createdF.flatMap { _ =>
+      isInitialized()
+    }
     isInitF.flatMap { isInit =>
       if (isInit) {
         FutureUtil.unit
       } else {
-        val createdF = ChainDbManagement.createAll()(this, ec)
         val genesisHeader =
           BlockHeaderDbHelper.fromBlockHeader(height = 0,
                                               bh =

--- a/db-commons/src/main/scala/org/bitcoins/db/DbManagement.scala
+++ b/db-commons/src/main/scala/org/bitcoins/db/DbManagement.scala
@@ -29,7 +29,14 @@ abstract class DbManagement extends DatabaseLogger {
 
     val query = {
       val querySeq =
-        allTables.map(createTableQuery(_, createIfNotExists = true))
+        allTables
+          .map(createTableQuery(_, createIfNotExists = true))
+          .map { query =>
+            // DIRTY HACK. For some reason Slick doesn't know that Sqlite can do CREATE INDEX IF NOT EXISTS
+            val statements = query.statements.map(
+              _.replace("create index", "create index if not exists"))
+            query.overrideStatements(statements)
+          }
       DBIO.seq(querySeq: _*).transactionally
     }
 

